WIP: Stream results rows to result set instead of buffering them all in memory

gmokki opened the pull request:
This is a WIP PR to gather opinions if this is a good direction

Currently pgjdbc by default buffers full ResultSet in memory before returning it to application.
This patch tries with minimal changes modify the implementation to return ResultSet already after the first row is received and then process more rows when ResultSet next() is called. This reduces the memory usage and lowers the latency.

The setFetchSize can currently be used to create a server side cursor to get a similar behavior. But that actually changes the query plan on the server side to assume that only first few items are requested. A cursor also requires extra round-trips to request more rows.


### All Submissions:

* [:heavy_check_mark:] Have you followed the guidelines in our [Contributing](https://github.com/pgjdbc/pgjdbc/blob/master/CONTRIBUTING.md) document?
* [:heavy_check_mark:] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### New Feature Submissions:

1. [:no_entry_sign:] Does your submission pass tests? No, but many happy day scenarios already work.
2. [:no_entry_sign:] Does mvn checkstyle:check pass ?
3. [:heavy_check_mark:] Have you added your new test classes to an existing test suite in alphabetical order?

### Changes to Existing Features:

* [:heavy_check_mark:] Does this break existing behaviour? Yes. An application written against older pgjdbc version might assume that ResultSet.next() will never throw SQLException, but with streamed results that is possible. A configuration option can be used to enable old behavior.
* [:heavy_check_mark:] Have you added an explanation of what your changes do and why you'd like us to include them?
* [:heavy_check_mark:] Have you written new tests for your core changes, as applicable?
* [:no_entry_sign:] Have you successfully run tests with your changes locally?

davecramer commented on docs/documentation/head/connect.md:
```diff
@@ -229,6 +229,15 @@ Connection conn = DriverManager.getConnection(url);
 
+* **streamResults** = boolean
+
+    since 42.2.12
```
So this should probably. be 43.3.0. It is a breaking change as are a number of others coming up

gmokki replied:
Yeah. this is still WIP. So let's keep this comment visible here until we start considering merging and then fix the number.

davecramer replied:
Fair enough. Also have a look at @vlsi gradle PR too. It will likely be merged in before this.

davecramer commented on pgjdbc/src/main/java/org/postgresql/core/BaseConnection.java:
```diff
@@ -212,4 +212,12 @@ CachedQuery createQuery(String sql, boolean escapeProcessing, boolean isParamete
   boolean hintReadOnly();
+
+  /**
+   * Prefer streaming ResultSet rows instead of buffring them.
```
s/buffring/buffering

ebrandsberg commented:
I have asked my engineer adrklos to help review and monitor this task, and to provide feedback, as this is a task that is of high interest to us as well (for Heimdall Data).  Fell free to contact him if you wish to have another person review any code.  As a side-note, we were able to avoid the majority of the performance overhead of the cursor-based approach by implementing a multi-threaded processor that reads each block of data based on an adaptive fetch size (https://github.com/pgjdbc/pgjdbc/pull/1718), and then writes the data to the upstream layer for use as needed.  The majority of the delay is waiting for the next block of data, and if you can incur this time while processing the previous block, most delays are avoided.

gmokki commented:
Thank you, reviews and suggestions how to make the code more clean are very welcome.
There was still one test that randomly failed, but other than that I think the code is ready.

I tried to benchmark this between my laptop and home server but really could not get much difference in speed. Which is also a good thing since it proves this does not slow down the normal cases.

This PR should hopefully fix both problematic cases: running out of memory by default or the extra waits when using fetch size.
It took a while to get this to stable state since the jdbc spec says that one can have multiple results sets open at the same time in one connection. In that case I just do the previous buffer everything in memory.
This can lead to situations where on first invocation when the code processes a resultset and asks for metadata, the metadata api triggers an extra query and forces the original query to be buffered in memory. The pgjdbc does cache the metadata at least for prepared statements so subsequent calls should work better.

-Mikko
________________________________
From: ebrandsberg <notifications@github.com>
Sent: 16 April 2020 17:35
To: pgjdbc/pgjdbc <pgjdbc@noreply.github.com>
Cc: Mikko Tiihonen <mikko.tiihonen@nitor.com>; Author <author@noreply.github.com>
Subject: Re: [pgjdbc/pgjdbc] WIP: Stream results rows to result set instead of buffering them all in memory (#1735)


I have asked my engineer adrklos to help review and monitor this task, and to provide feedback, as this is a task that is of high interest to us as well (for Heimdall Data). Fell free to contact him if you wish to have another person review any code. As a side-note, we were able to avoid the majority of the performance overhead of the cursor-based approach by implementing a multi-threaded processor that reads each block of data based on an adaptive fetch size (#1718<https://github.com/pgjdbc/pgjdbc/pull/1718>), and then writes the data to the upstream layer for use as needed. The majority of the delay is waiting for the next block of data, and if you can incur this time while processing the previous block, most delays are avoided.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/pgjdbc/pgjdbc/pull/1735#issuecomment-614692504>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AABJ2I233Z6MVEYQD5HVLFTRM4JT5ANCNFSM4LK472WA>.

adrklos commented:
The idea of streaming result rows look really good, especially when it helps to avoid OutOfMemory exception during reading big results. The one thing I see somehow "ignored" is maxResultBuffer property (I don't see this as intentional, the property was added not that long ago). The purpose of this property is to declare max memory size which can be used during one reading of a result.  I see two scenarios when some changes may be required when streamResults and maxResultBuffer is used (what gonna happen when only maxResultBuffer is declared, because by now streamResults by default is true):

1. MaxResultBuffer is "overprotective", when next() is called to process next row. So far size of read result is computed when rows of result are processed and cleared when "Ready for query" packet is read. With streamResults it should be cleared after reading streamed row, because there gonna be no reading of next rows, so that's all of memory which was used.
2. MaxResultBuffer isn't "protective" enough, when new resultSet is created. In this situation, previous resultSet is buffered and a row of new resultSet is read, but with streamResult and maxResultBuffer, count of read bytes is cleared after reading 'Z' packet, so size of a row from the new result isn't really counted, what may lead to exceeding declared maxResultBuffer and maybe to OutOfMemory exception if row was big enough. Again it looks like the best would be to clear size of read result after reading a row of streamed resultSet.

So for both scenarios would be the best to clear size of read result after read of a row of streamed resultSet and turn it off when "Ready for query" ('Z') packet is read if streamResult is true. For better understanding of work of counting bytes of read result I propose to see fields `maxResultBuffer` and `resultBufferByteCount` and methods `receiveTupleV3()` , `increaseByteCounter(long value)` and `clearResultBufferCount()` in [PGStream](https://github.com/gmokki/pgjdbc/blob/stream-results/pgjdbc/src/main/java/org/postgresql/core/PGStream.java) class and how they are used in the code.

For added tests I would want to see a test, which checks if streaming was truly used, because I don't see in the current tests any part which checks this (correct me if I'm wrong). Then maybe a test which checks if streaming was NOT used when streamResult is false. And maybe a test to check if buffering was used/done if new ResultSet is created and previous ResultSet should be buffered. The purpose of proposed tests would be to check if code really does what we think it's doing.

gmokki commented:
I can have a go at  maintaining the resultBufferByteCount correctly when streaming (and switching modes when streaming is changed back to buffering mode while reading the result set). Thanks for the intro.

I did add a StreamResultSetTest that verifies that streaming is done correctly. I did it by adding a test that pg_sleeps in backend between every result set row. And then I verify that time difference between the result set first row and last row is longer than sleep time. If the streaming is disabled then the result set is buffered (which takes n*sleep) and code processes each row in zero time.

But I have not tested that the streaming=false still works.

-Mikko

________________________________
From: adrklos <notifications@github.com>
Sent: 21 April 2020 12:59
To: pgjdbc/pgjdbc <pgjdbc@noreply.github.com>
Cc: Mikko Tiihonen <mikko.tiihonen@nitor.com>; Author <author@noreply.github.com>
Subject: Re: [pgjdbc/pgjdbc] WIP: Stream results rows to result set instead of buffering them all in memory (#1735)


The idea of streaming result rows look really good, especially when it helps to avoid OutOfMemory exception during reading big results. The one thing I see somehow "ignored" is maxResultBuffer property (I don't see this as intentional, the property was added not that long ago). The purpose of this property is to declare max memory size which can be used during one reading of a result. I see two scenarios when some changes may be required when streamResults and maxResultBuffer is used (what gonna happen when only maxResultBuffer is declared, because by now streamResults by default is true):

  1.  MaxResultBuffer is "overprotective", when next() is called to process next row. So far size of read result is computed when rows of result are processed and cleared when "Ready for query" packet is read. With streamResults it should be cleared after reading streamed row, because there gonna be no reading of next rows, so that's all of memory which was used.
  2.  MaxResultBuffer isn't "protective" enough, when new resultSet is created. In this situation, previous resultSet is buffered and a row of new resultSet is read, but with streamResult and maxResultBuffer, count of read bytes is cleared after reading 'Z' packet, so size of a row from the new result isn't really counted, what may lead to exceeding declared maxResultBuffer and maybe to OutOfMemory exception if row was big enough. Again it looks like the best would be to clear size of read result after reading a row of streamed resultSet.

So for both scenarios would be the best to clear size of read result after read of a row of streamed resultSet and turn it off when "Ready for query" ('Z') packet is read if streamResult is true. For better understanding of work of counting bytes of read result I propose to see fields maxResultBuffer and resultBufferByteCount and methods receiveTupleV3() , increaseByteCounter(long value) and clearResultBufferCount() in PGStream<https://github.com/gmokki/pgjdbc/blob/stream-results/pgjdbc/src/main/java/org/postgresql/core/PGStream.java> class and how they are used in the code.

For added tests I would want to see a test, which checks if streaming was truly used, because I don't see in the current tests any part which checks this (correct me if I'm wrong). Then maybe a test which checks if streaming was NOT used when streamResult is false. And maybe a test to check if buffering was used/done if new ResultSet is created and previous ResultSet should be buffered. The purpose of proposed tests would be to check if code really does what we think it's doing.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/pgjdbc/pgjdbc/pull/1735#issuecomment-617080797>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AABJ2I2TESCHH377CH2ZWWLRNVVATANCNFSM4LK472WA>.

adrklos commented:
I don't feel completely convinced to that way of testing this, but after analysing again these tests I see that they are somehow proving that streaming is done. Still I have two concerns:

1. I feel like streaming results is made to improve memory usage and changes in memory should be used to prove that (but on the other hand to don't go too far with it, in my opinion i.e. it should be enough to see changes in object of StreamingList or DynamicListIterator).
2. (Please, correct me if I did something wrong here) I ran one of your tests (`testStreamedResults()`), but I changed sleep time from 0.001 s to 0.1s which gave me 100s of waiting for result (which I applied). I noticed that database awaits for rows as long as they can be contained in the packet to send, but if next data row will exceed packet size, then packet is send to application and new packet is begin created by database. During my test one packet could contain 457 data rows, so first row I got after 45.7 seconds from the start, after 100 seconds from the start I got last data row, so the difference was near 54 seconds. It proves that first row was read faster than last (and it's really smart way to test that), but checking difference between first and last row isn't convincing (in your situation, time between first and last row should be near 0.54 s, so the rest of time required to pass the test is added by code work). Additionally, would be good to add more explanation why these tests are made like that, especially to explain how sleeping proves reading first row before last row and all restrictions required by these tests (i.e. with smaller number of rows all data would be send in one packet and all rows would be read at once).

adrklos commented on pgjdbc/src/test/java/org/postgresql/test/jdbc4/StreamResultSetTest.java:
```diff
@@ -0,0 +1,110 @@
+    // the streaming features works nicely though
+    assumeTrue(haveMinimumServerVersion(con, v10));
+
+    try (ResultSet results = statement.executeQuery()) {
+      int count = 0;
+      long startTime = currentTimeMillis();
+      long fistRowTime = 0;
+      while (results.next()) {
+        int value = results.getInt(1);
+        if (fistRowTime == 0) {
+          fistRowTime = currentTimeMillis();
+        }
```
Isn't reading firstRowTime and startTime make them always "same" in this situation? ResultSet have to read first row with or without streaming, so coming to this point is really fast, because all waiting for first row is made in executeQuery() method. The only situation when it doesn't happen is when resultSet is empty, but it isn't tested in this situation. I think that startTime should be set before line 91 to really check if data packet with first row was read without waiting for data packet with last row.

davecramer replied:
No it's after the first row has been read. Which could be different. Also `fistRow` is a spelling mistake

adrklos replied:
But `executeQuery()` reads first row right? So if `startTime` and `firstRow` both are read after `executeQuery()` then their times are really close, `next()` doesn't affect reading first row in this situation I think.

gmokki replied:
Fixed the typo in variable and moved the startTime before the executeQuery.

gmokki replied:
I rewrote the streaming test. Now it has less hard coded random values and lots of comments explaining what is being tested. I also verified that the server seems to buffer results into 8kB blocks, so countered that by adding some padding to the rows to force them in multiple packets.

davecramer commented:
@gmokki Where are we with this ?

gmokki commented:
I added clearing the `resultBufferByteCount` when in streaming mode (and not reverted back to buffering mode).
Also managed to fix the final randomly failing test `StatementTest.testFastCloses` caused by locking bug.

davecramer commented:
@gmokki Hi Mikko, any more progress on this? Also check out https://github.com/pgjdbc/pgjdbc/pull/1789 to see how this PR effects yours?

gmokki commented:
I'm sorry for not pushing this through earlier.

I rebased and resolved all conflicts with the adaptive streaming and other changes. Added required checker annotations to make the tests pass.
Also rewrote the streaming test that previously only worked on local machine to also work on slower travis test machines.
Move some code back to java8 instead of using inner classes.

davecramer commented:
@vlsi can you have a look at this?

vlsi commented on pgjdbc/src/main/java/org/postgresql/core/v3/QueryExecutorImpl.java:
```diff
@@ -2063,29 +2102,154 @@ private void processDeadPortals() throws IOException {
+  }
+
+  private @Nullable Tuple processStreamedResultsImpl(ProcessState state) throws SQLRuntimeException {
+    synchronized (this) {
```
This likely goes away from the recent effort of removing `synchonized` for better Loom compatibility

davecramer replied:
I didn't think we pushed that yet. I was going to put the loom stuff in an experimental branch until it's all sorted.

vlsi commented:
@gmokki , do you have bechmarks for "before vs after" the change?

In other words:
4.3.0
4.3.0 + patch, stream=off
4.3.0 + patch, stream=on

davecramer commented:
This should be interesting. I expect the difference will be seen in large result sets

gmokki commented:
So far I've been unable to get any measurable difference when testing against local postgresql.
What kind of benchmarks would be good to run?
* simple `select 1` test to see the overhead
* 100k rows of 1k json objects to test the actual streaming

vlsi commented:
>simple select 1 test to see the overhead

+1

>100k rows of 1k json objects to test the actual streaming

+1

+ "autocommit=false, 100k rows of 1k json objects to test the actual streaming".
This should distinguish "fetch all rows" vs "fetch by 1000 rows" cases.

davecramer commented:
> So far I've been unable to get any measurable difference when testing against local postgresql.


Ya it's difficult on localhost to get any difference. Are there any proxies out there that introduce lag?

ebrandsberg commented:
Honestly, performance is the least concern about a fix like this.  It is about not overflowing memory.  My engineers are working on other tasks, but I will assign them to test this in our testbed and see if they can see any performance deltas of significance, although it may be next week before we can get to it.

gmokki commented:
You are correct that performance improvement is not what we're after. But we don't want to regress too much on the normal flow if we want this to be enabled by default.

There are two cases where I can see performance is improved:
- OOM vs not -> huge difference since JVM starts to slow down when getting close to OOM
- setFetchSize vs just streaming when fetching all rows -> fetch size has many extra round-trips and suboptimal plan on backend

If you test this PR do note that many ResultSet methods seem to trigger a backend query and thus require the result set to be buffered anyway into memory. For this to get any real use we should see if we could run them somehow earlier, for example for prepared statements and not wait for the first hit to metadata to fetch for example column names (I haven't yet checked if that triggers the buffering, but I suspect it does).

ebrandsberg commented:
> You are correct that performance improvement is not what we're after. But we don't want to regress too much on the normal flow if we want this to be enabled by default.
> 
> There are two cases where I can see performance is improved:
> 
> * OOM vs not -> huge difference since JVM starts to slow down when getting close to OOM
> * setFetchSize vs just streaming when fetching all rows -> fetch size has many extra round-trips and suboptimal plan on backend
> 
> If you test this PR do note that many ResultSet methods seem to trigger a backend query and thus require the result set to be buffered anyway into memory. For this to get any real use we should see if we could run them somehow earlier, for example for prepared statements and not wait for the first hit to metadata to fetch for example column names (I haven't yet checked if that triggers the buffering, but I suspect it does).

Can I assume that when it reverts to the buffer, the maxResultBuffer is being honored still?  We added that code, and have submitted the same feature with the same behavior to the SQL Server and Oracle MySQL driver for consistency.

gmokki commented:
I at least tried to make it honor the maxResultBuffer and some auto tuned fetch size options. It is really using almost exactly same code, but instead of adding the parsed Tuple to an array the protocol reader instead returns back immediately.

vlsi commented:
I've rebased the PR, let us see if it still works.

vlsi commented:
@gmokki , I see you add hamcrest to `reduced-pom.xml`, however, I'm afraid that does not work. It would require `hamcrest` to be packaged at the build machine which does not seem to be the case.

I'm inclined that we either skip `hamcrest` use in the newly added code, or we could just skip the newly added test in `reduced-pom.xml` execution.

All the `reduced-pom.xml` dependencies must be present in https://github.com/pgjdbc/pgjdbc/blob/f61fbfe7b72ccf2ca0ac2e2c366230fdb93260e5/packaging/rpm/postgresql-jdbc.spec.tpl#L52-L63, so adding new dependencies is hard (one needs to verify if the dependency is packaged :-( )

vlsi commented:
@gmokki , sorry for being late here, however, it would really help if you could remember the reason you introduced `StreamingList`. Was it to minimize `List<Tuple>` changes? Frankly speaking, I am a bit scared of adding `List` and `ListIterator` implementation without test coverage. I would rather replace `List<Tuple>` with something less invasive (e.g. our own list-like interface with a minimal set of methods)

vlsi commented:
I've pushed a basic implementation of `TupleBuffer`. What I did was I created an empty interface, then I asked IDEA to migrate all uses of `List<Tuple>` to the new interface. Then I added methods to get the code to compile.

Now I plan to remove all references of `org.postgresql.jdbc.tuple.TupleBuffer#size()` since the method does not play with lazy loading well.

Sample issue:


```java
class PgResultSet {
  public boolean isAfterLast() throws SQLException {
...
    final int rows_size = rows.size();
    if (rowOffset + rows_size == 0) {
      return false;
    }
...
```

The issue is that `rows` can point to a half-loaded buffer, and we probably need to call `FETCH` to ensure there are no more rows.

---

Frankly speaking, I don't quite like the separation between `PgResultSet`, `TupleBuffer`, `Provider<Tuple>`, and `Cursor`. It is not clear whose responsibility is to send `fetch` requests to the database. Currently it is done in `PgResultSet`, however, it makes `TupleBuffer` (~ `StreamingList`) awkward since the lazy list can't tell between "all the rows are fetched" vs "the recently requested rows exhausted and one more fetch command can be sent to the DB".

ebrandsberg commented:
Any more work on this?  I feel this is important for BI tools to be able to handle arbitrary results without having to wrap them in transactions, and streaming logic like this is important.  It looks like this has gone stale for over a year however.  Thanks!

ebrandsberg commented:
For anybody looking at this, we will be reviving a PR for this, having worked to test and validate the fix.  It appears there was at least one issue that resulted in OOM with the change active, and our engineering team has fixed it.

Chris-SP365 commented:
In our applications, one can export/import the databases. The columns can be configured inside the app, too. They can contain 1-x blob columns.

If we set fetch size to 1, it is slow (specially for tables with a lot of entries).
If we set fetch size to > 1 we don't know if it fits in memory.
If we set adaptive fetch we don't know if it fits in memory.
The less resources, the better.

SQL Server defaults to this:
https://learn.microsoft.com/en-us/sql/connect/jdbc/using-adaptive-buffering?view=sql-server-ver16
MySQL has the option to set it via setFetchSize (Integer.MIN_VALUE), see Chapter 'ResultSet':
https://dev.mysql.com/doc/connector-j/en/connector-j-reference-implementation-notes.html

So is there anything which has to be resolved for this PR? Or is there a reason that this will not be included into pgjdbc at all?
